src/home_robot/home_robot/mapping/voxel.py

```diff
@@ -25,88 +53,470 @@ def combine_point_clouds(
-        assert xyz.shape[-1] == 3
-        assert feats.shape[-1] == self.feature_dim
-        assert xyz.shape[0] == feats.shape[0]
-        self.observations.append((camera_pose, xyz, feats, info))
-        world_xyz = trimesh.transform_points(xyz, camera_pose)
+    def get_instances(self) -> List[InstanceView]:
+        """Return a list of all viewable instances"""
+        return self.instances.instance_views[0].values()
+
+    def add(
+        self,
+        camera_pose: np.ndarray,
+        xyz: np.ndarray,
+        rgb: np.ndarray,
+        feats: Optional[np.ndarray] = None,
+        depth: Optional[np.ndarray] = None,
+        base_pose: Optional[np.ndarray] = None,
+        obs: Optional[Observations] = None,
+        **info,
+    ):
+        """Add this to our history of observations. Also update the current running map.
+
+        Parameters:
+            camera_pose(np.ndarray): necessary for measuring where the recording was taken
+            xyz: N x 3 point cloud points
+            rgb: N x 3 color points
+            feats: N x D point cloud features; D == 3 for RGB is most common
+            base_pose: optional location of robot base
+            obs: observations
+        """
+        # TODO: we should remove the xyz/feats maybe? just use observations as input?
+        # TODO: switch to using just Obs struct?
+        assert (
+            xyz.shape[-1] == 3
+        ), "xyz must have last dimension = 3 for x, y, z position of points"
+        W, H, _ = xyz.shape
+        assert rgb.shape == xyz.shape, "rgb shape must match xyz"
+        if feats is not None:
+            assert (
+                feats.shape[-1] == self.feature_dim
+            ), f"features must match voxel feature dimenstionality of {self.feature_dim}"
+            assert (
+                xyz.shape[0] == feats.shape[0]
+            ), "features must be available for each point"
+        else:
+            pass
+
+        # add observations before we start changing things
+        self.observations.append(
+            Frame(camera_pose, xyz, rgb, feats, depth, base_pose, obs, info)
+        )
+
+        if feats is not None:
+            feats = feats.reshape(-1, feats.shape[-1])
+        rgb = rgb.reshape(-1, 3)
+        xyz = xyz.reshape(-1, 3)
+        full_world_xyz = trimesh.transform_points(xyz, camera_pose)
 
+        if depth is not None:
+            # Apply depth filter
+            valid_depth = np.bitwise_and(depth > self.min_depth, depth < self.max_depth)
+            valid_depth = valid_depth.reshape(-1)
+            if feats is not None:
+                feats = feats[valid_depth, :]
+            rgb = rgb[valid_depth, :]
+            xyz = xyz[valid_depth, :]
+            world_xyz = full_world_xyz[valid_depth, :]
+        else:
+            valid_depth = None
+
+        # TODO: tensorize earlier
+        # Note that in process_instances_for_env, we assume that the background class is 0
```

Comment:
Can we add a param `background_instance_id: int` to process_instances_for_env? This removes a hard-coded/hidden parameter in that function, and saves us from having to work around it here :)

Reply:
done